File: 백준/Gold/16236. 아기 상어/아기 상어.cc
#include<iostream>
#include<algorithm>
#include<vector>
#include<queue>
#include<utility>
#include<cstring>
using namespace std;
struct Fish {
	int s, x, y;
};
typedef pair<int, int>p;
p now;
int n, ans, sharkSize = 2, expe;
int arr[21][21], check[21][21];
int dx[4] = { 0,0,-1,1 }, dy[4] = { 1,-1,0,0 };
vector<Fish>canEat;
queue<p>q;
bool compare(const Fish& f1, const Fish& f2) {
	if (f1.s == f2.s) {
		if (f1.x == f2.x) return f1.y < f2.y;
		return f1.x < f2.x;
	}
	return f1.s < f2.s;
}
bool inValid(int x, int y) {
	return !(x<0 || x>n - 1 || y<0 || y>n - 1);
}
void input() {
	cin >> n;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> arr[i][j];
			if (arr[i][j] == 9) {
				now = { i,j };
				arr[i][j] = 0;
			}
		}
	}
	expe = 2;
}
int bfs() {
	q.push(now);
	while (!q.empty()) {
		int x = q.front().first, y = q.front().second;
		q.pop();
		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i], ny = y + dy[i];
			if (!inValid(nx,ny)|| arr[nx][ny] > sharkSize||check[nx][ny])continue; 
			check[nx][ny] = check[x][y] + 1;
			q.push({ nx,ny });
			if (arr[nx][ny] && arr[nx][ny] < sharkSize) {
				canEat.push_back({ check[nx][ny],nx,ny });
			}
		}
	}
	return canEat.size();
}
void eat(Fish target) {
	ans += check[target.x][target.y];
	arr[target.x][target.y] = 0;
	now = { target.x,target.y };
	expe--;
	if (!expe) {
		expe = ++sharkSize;
	}
}
int main() {
	ios_base::sync_with_stdio(false), cin.tie(0), cout.tie(0);
	input();
	while (bfs()) {
		sort(canEat.begin(), canEat.end(), compare);
		Fish target = canEat[0];
		eat(target);
		memset(check, 0, sizeof(check));
		canEat.clear();
	}
	cout << ans << "\n";
}

File: 백준/Gold/16236. 아기 상어/아기 상어.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static class Fish implements Comparable<Fish> {
		int s, x, y;

		public Fish(int s, int x, int y) {
			super();
			this.s = s;
			this.x = x;
			this.y = y;
		}

		@Override
		public int compareTo(Fish o) {
			if (this.s == o.s) {
				if (this.x == o.x)
					return this.y - o.y;
				return this.x - o.x;
			}
			return this.s - o.s;
		}
	}

	static class Pos {
		int x, y;

		public Pos(int x, int y) {
			super();
			this.x = x;
			this.y = y;
		}

	}

	static int n, ans, sharkSize = 2, expe = 2;
	static int[][] arr, check;
	static int[] dx = { 0, 0, -1, 1 }, dy = { 1, -1, 0, 0 };
	static Pos now;
	static List<Fish> canEat;
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static Queue<Pos> q = new LinkedList<>();
	static StringTokenizer st;

	static boolean inValid(int x, int y) {
		return !(x < 0 || x > n - 1 || y < 0 || y > n - 1);
	}

	static void input() throws NumberFormatException, IOException {
		n = Integer.parseInt(br.readLine());
		arr = new int[n][n];
		check = new int[n][n];
		canEat=new ArrayList<>();
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < n; j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
				if (arr[i][j] == 9) {
					now = new Pos(i, j);
					arr[i][j] = 0;
				}
			}
		}
	}

	static int bfs() {
		q.add(now);
		while (!q.isEmpty()) {
			int x = q.peek().x, y = q.poll().y;
			for (int i = 0; i < 4; i++) {
				int nx = x + dx[i], ny = y + dy[i];
				if (!inValid(nx, ny) || arr[nx][ny] > sharkSize || check[nx][ny] > 0)
					continue;
				check[nx][ny] = check[x][y] + 1;
				q.add(new Pos(nx, ny));
				if (arr[nx][ny] > 0 && arr[nx][ny] < sharkSize) {
					canEat.add(new Fish(check[nx][ny], nx, ny));
				}
			}
		}
		return canEat.size();
	}

	static void eat(Fish target) {
		ans += check[target.x][target.y];
		arr[target.x][target.y] = 0;
		now.x = target.x;
		now.y = target.y;
		expe--;
		if (expe == 0) {
			expe = ++sharkSize;
		}
	}

	public static void main(String[] args) throws NumberFormatException, IOException {
		input();
		while (bfs() > 0) {
			Collections.sort(canEat);
			Fish target = canEat.get(0);
			eat(target);
			for(int[] i: check)Arrays.fill(i, 0);
			canEat.clear();
		}
		System.out.println(ans);
	}

}